speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -1,4 +1,134 @@
+                                 const std::string& p_environmentName,
+                                 int p_environmentVersion,
+                                 InterventionType p_interventionType
+) {
+    m_saveSettings = p_saveSettings;
+    m_outputStream.open(p_fileName);
+    // User and trial data
+    WRITE_STRING_VAR(m_outputStream, p_userId);
+    writeTime(m_outputStream, p_trialStartTime);
+    // Black box data
+    WRITE_STRING_VAR(m_outputStream, p_blackboxFilename);
+    writeTime(m_outputStream, p_blackboxTime);
+    WRITE_STRING_VAR(m_outputStream, p_blackboxName);
+    // Environment data
+    WRITE_STRING_VAR(m_outputStream, p_environmentFilename);
+    WRITE_VAR(m_outputStream, p_environmentVersion);
+    WRITE_STRING_VAR(m_outputStream, p_environmentName);
+    // Intervention data
+    WRITE_VAR(m_outputStream, p_interventionType);
+    // Headers to indicate what data will be saved
+    if (m_saveSettings.EnvironmentData) {
+        WRITE_STRING_LIT(m_outputStream, "GameState");
+    }
+    if (m_saveSettings.HumanData) {
+        WRITE_STRING_LIT(m_outputStream, "UserInput");
+    }
+}
+
+/// @brief Shutdown the file data storage.
+/// End result: any possible final data is written and the file is released.
+void FileDataStorage::Shutdown() {
+    m_outputStream << "END";
+    m_outputStream.close();
+}
+
+/// @brief Save the current driving situation to the buffer
+/// @param p_car Current car status in Speed Dreams
+/// @param p_situation Current situation in Speed Dreams
+/// @param p_timestamp Current tick
+void FileDataStorage::Save(tCarElt* p_car, tSituation* p_situation, unsigned long p_timestamp) {
+    WRITE_VAR(m_outputStream, p_timestamp);
+    if (m_saveSettings.EnvironmentData) {
+        Posd pos = p_car->pub.DynGCg.pos;
+        tDynPt mov = p_car->pub.DynGC;
+        WRITE_VAR(m_outputStream, pos.x); // x-position
+        WRITE_VAR(m_outputStream, pos.y); // y-position
+        WRITE_VAR(m_outputStream, pos.z); // z-position
+        WRITE_VAR(m_outputStream, pos.ax);// x-direction
+        WRITE_VAR(m_outputStream, pos.ay);// y-direction
+        WRITE_VAR(m_outputStream, pos.az);// z-direction
+        WRITE_VAR(m_outputStream, mov.vel.x); // speed
+        WRITE_VAR(m_outputStream, mov.acc.x); // acceleration
+        WRITE_VAR(m_outputStream, p_car->priv.gear); // gear
+    }
+    if (m_saveSettings.HumanData) {
+        tCarCtrl ctrl = p_car->ctrl;
+        WRITE_VAR(m_outputStream, ctrl.steer); // steer
+        WRITE_VAR(m_outputStream, ctrl.brakeCmd); // brake
+        WRITE_VAR(m_outputStream, ctrl.accelCmd); // gas
+        WRITE_VAR(m_outputStream, ctrl.clutchCmd); // clutch
+    }
+}
+
+/// @brief Save all decisions that were taken this tick
+/// @param p_decisions Tuple of decisions taken this tick
+void FileDataStorage::SaveDecisions(DecisionTuple& p_decisions) {
+    if (!m_saveSettings.InterventionData) return;
+    WRITE_STRING_LIT(m_outputStream, "Decisions");
```

Comment:
A new line here would make it more clear
```suggestion
    if (!m_saveSettings.InterventionData) return;
    
    WRITE_STRING_LIT(m_outputStream, "Decisions");
```